spec/lib/nerve/reporter_zookeeper_spec.rb

```diff
@@ -29,5 +32,109 @@
     reporter.report_up
     reporter.stop
   end
+
+  context "when reporter is up" do
+    before(:each) do
+      allow(zk).to receive(:close!)
+      allow(zk).to receive(:connected?).and_return(true)
+      allow(zk).to receive(:exists?) { "zk_path" }.and_return(false)
+      allow(zk).to receive(:mkdir_p) { "zk_path" }
+      allow(zk).to receive(:create) { "full_path" }
+      allow(zk).to receive(:set) { "full_path" }
+      allow(zk).to receive(:delete).with("full_path", anything())
+      allow(ZK).to receive(:new).and_return(zk)
+      @reporter = Nerve::Reporter::Zookeeper.new(subject)
+      @reporter.start
+      @reporter.report_up
+    end
+
+    after(:each) do
+      # reset the class variable to avoid mock object zk leak
+      Nerve::Reporter::Zookeeper.class_variable_set(:@@zk_pool, {})
+      Nerve::Reporter::Zookeeper.class_variable_set(:@@zk_pool_count, {})
+    end
+
+    it "returns true on report_up" do
+      expect(@reporter.report_up).to be true
+    end
+
+    it "returns true on report_down" do
+      expect(@reporter.report_down).to be true
+    end
+
+    it "returns true on ping?" do
+      expect(zk).to receive(:exists?) { "zk_path" }.and_return(true)
+      expect(@reporter.ping?).to be true
+    end
+
+    context "when zk.connected? started to return false" do
+      it 'returns false on ping?' do
+        # this condition is triggered if connection has been lost for a while (a few sec)
+        expect(zk).to receive(:connected?).and_return(false)
+        expect(zk).not_to receive(:exists?)
+        expect(@reporter.ping?).to be false
+      end
+
+      it 'returns false on report_up without zk operation' do
+        # this condition is triggered if connection has been lost for a while (a few sec)
+        expect(zk).to receive(:connected?).and_return(false)
+        expect(zk).not_to receive(:set)
+        expect(@reporter.report_up).to be false
+      end
+
+      it 'returns false on report_up without zk operation' do
+        # this condition is triggered if connection has been lost for a while (a few sec)
+        expect(zk).to receive(:connected?).and_return(false)
+        expect(zk).not_to receive(:delete)
+        expect(@reporter.report_down).to be false
+      end
+    end
+
+    context "when there is a short disconnection" do
+      it 'returns false on ping?' do
+        # this condition is triggered if connection is shortly interrupted
+        # so connected? still return true
+        expect(zk).to receive(:exists?).and_raise(ZK::Exceptions::OperationTimeOut)
```

Comment:
likewise this should live in the context, not in the example blocks

Reply:
for report_up and report_down, they actually throw exception from a different zk operation, not zk.exists?. so there is no need to have a common before block